pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/SegmentMergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,199 @@
+      // based on the segment lineage.
+      SegmentLineage segmentLineageForTable = _clusterInfoProvider.getSegmentLineage(offlineTableName);
+      Set<String> segmentsNotToMerge = new HashSet<>();
+      if (segmentLineageForTable != null) {
+        for (String segmentLineageEntryId : segmentLineageForTable.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineageForTable.getLineageEntry(segmentLineageEntryId);
+          // Segments shows up on "segmentFrom" field in the lineage entry should not be scheduled again.
+          segmentsNotToMerge.addAll(lineageEntry.getSegmentsFrom());
+
+          // Segments shows up on "segmentsTo" field in the lineage entry with "IN_PROGRESS" state cannot be merged.
+          if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+            segmentsNotToMerge.addAll(lineageEntry.getSegmentsTo());
+          }
+        }
+      }
+
+      // Filter out the segments that cannot be merged
+      List<SegmentZKMetadata> segmentsToMergeForTable = new ArrayList<>();
+      List<String> scheduledSegments = scheduledSegmentsMap.getOrDefault(offlineTableName, Collections.emptyList());
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : segmentsForOfflineTable) {
+        String segmentName = offlineSegmentZKMetadata.getSegmentName();
+
+        // The segment should not be merged if it's already scheduled or in progress
+        if (scheduledSegments.contains(segmentName) || segmentsNotToMerge.contains(segmentName)) {
+          continue;
+        }
+
+        // The segments that are newer than the buffer time period should not be be merged
+        if (System.currentTimeMillis() - offlineSegmentZKMetadata.getEndTime() < bufferTimePeriodMs) {
+          continue;
+        }
+        segmentsToMergeForTable.add(offlineSegmentZKMetadata);
+      }
+
+      // Compute Merge Strategy
+      List<List<SegmentZKMetadata>> segmentsToSchedule = MergeStrategyFactory.getMergeStrategy(taskConfigs)
+          .generateMergeTaskCandidates(segmentsToMergeForTable, maxNumSegmentsPerTask);
+
+      // Generate tasks
+      for (List<SegmentZKMetadata> segments : segmentsToSchedule) {
+        if (tableNumTasks == tableMaxNumTasks) {
+          break;
+        }
+        if (segments.size() >= 1) {
+          Map<String, String> configs = new HashMap<>();
+          configs.put(MinionConstants.TABLE_NAME_KEY, offlineTableName);
+          configs.put(MinionConstants.SEGMENT_NAME_KEY,
+              segments.stream().map(s -> s.getSegmentName()).collect(Collectors.joining(",")));
+          configs.put(MinionConstants.DOWNLOAD_URL_KEY,
+              segments.stream().map(s -> ((OfflineSegmentZKMetadata) s).getDownloadUrl())
+                  .collect(Collectors.joining(",")));
+          configs.put(MinionConstants.VIP_URL_KEY, _clusterInfoProvider.getVipUrl());
+          configs.put(MinionConstants.REPLACE_SEGMENTS_KEY, Boolean.TRUE.toString());
```

Comment:
What is this config for?

Reply:
For multiple segments minion job, I made `replaceSegment` API to be optional. So, users can choose to use the atomic replacement of the segment or they choose not to.

Reply:
For this specific job type, I don't think we need this extra config. On the minion side, it should always use the `replaceSegment` API, instead of picking it from the task config.

Reply:
I added it because some minion tasks may not want to use replaceSegment API? (e.g. I'm not sure if we want to use replaceSegment for realtime to offline job)

Reply:
I don't fully understand. This generator is for `SegmentMergeRollupTask` only, and there will be an executor for this task type. The executor can set this flag to be true, but this should not be set by the generator.